dali/test/python/test_FW_iterators_shuffling.py

```diff
@@ -0,0 +1,314 @@
+                                    pad_last_batch=pad_last_batch, initial_fill=initial_fill)
+
+    def define_graph(self):
+        images, bb, labels, ids = self.input(name="Reader")
+        return ids
+
+test_data_root = os.environ['DALI_EXTRA_PATH']
+coco_folder = os.path.join(test_data_root, 'db', 'coco')
+data_sets = [[os.path.join(coco_folder, 'images'), os.path.join(coco_folder, 'instances.json')]]
+
+
+def test_mxnet_iterator_last_batch_no_pad_last_batch():
+    num_gpus = 1
+    batch_size = 100
+    iters = 0
+
+    #make sure that data size and batch are not divisible
+    while iters % batch_size == 0:
+        while iters != 0 and iters % batch_size == 0:
+            batch_size += 1
+
+        pipes = [COCOReaderPipeline(batch_size=batch_size, num_threads=4, shard_id=gpu, num_gpus=num_gpus,
+                                                        data_paths=data_sets[0], random_shuffle=True, stick_to_shard=False,
+                                                        shuffle_after_epoch=False, pad_last_batch=False) for gpu in range(num_gpus)]
+        [pipe.build() for pipe in pipes]
+        iters = pipes[0].epoch_size("Reader")
+        iters = iters // num_gpus
+
+    data_size = iters
+
+    dali_train_iter = MXNetIterator(pipes, [("ids", MXNetIterator.DATA_TAG)],
+                                    size=pipes[0].epoch_size("Reader"), fill_last_batch=True)
+
+    img_ids_list = []
+    for it in iter(dali_train_iter):
+        tmp = it[0].data[0].squeeze().asnumpy().copy()
+        img_ids_list.append(tmp)
+    img_ids_list = np.concatenate(img_ids_list)
+    img_ids_list_set = set(img_ids_list)
+
+    remainder = int(math.ceil(data_size / batch_size)) * batch_size - data_size
+    mirrored_data = img_ids_list[-remainder - 1:]
+
+    assert len(img_ids_list) > data_size
+    assert len(img_ids_list_set) == data_size
+    assert len(set(mirrored_data)) != 1
+
+
+def test_mxnet_iterator_last_batch_pad_last_batch():
+    num_gpus = 1
+    batch_size = 100
+    iters = 0
+
+    #make sure that data size and batch are not divisible
+    while iters % batch_size == 0:
+        while iters != 0 and iters % batch_size == 0:
+            batch_size += 1
+        pipes = [COCOReaderPipeline(batch_size=batch_size, num_threads=4, shard_id=gpu, num_gpus=num_gpus,
+                                                        data_paths=data_sets[0], random_shuffle=True, stick_to_shard=False,
+                                                        shuffle_after_epoch=False, pad_last_batch=True) for gpu in range(num_gpus)]
+        [pipe.build() for pipe in pipes]
+        iters = pipes[0].epoch_size("Reader")
+        iters = iters // num_gpus
+
+    data_size = iters
+
+    dali_train_iter = MXNetIterator(pipes, [("ids", MXNetIterator.DATA_TAG)],
+                                    size=pipes[0].epoch_size("Reader"), fill_last_batch=True)
+
```

Comment:
I see duplicated code with respect to previous test cases. Extract utilities?

Reply:
Done